module/zfs/vdev_draid.c

```diff
@@ -0,0 +1,1669 @@
+	permutation = kmem_alloc(sizeof (permutation[0]) * ncols, KM_SLEEP);
+	VERIFY0(vdev_draid_get_permutation(permutation, perm, cfg));
+
+	for (c = 0; c < mm->mm_children; c++) {
+		int idx = group * copies + c;
+		mirror_child_t *mc = &mm->mm_child[c];
+
+		/* The remainder group is not usable for IO */
+		ASSERT3U(idx, <, ((ncols - nspare) / copies) * copies);
+
+		mc->mc_vd = vd->vdev_child[permutation[idx]];
+		mc->mc_offset = o;
+	}
+
+	if (array == NULL)
+		kmem_free(permutation, sizeof (permutation[0]) * ncols);
+	else
+		*array = permutation; /* caller will free */
+
+	zio->io_vsd = mm;
+	zio->io_vsd_ops = &vdev_mirror_vsd_ops;
+	return (mm);
+}
+
+static inline void
+vdev_draid_assert_vd(const vdev_t *vd)
+{
+	struct vdev_draid_configuration *cfg = vd->vdev_tsd;
+
+	ASSERT3P(vd->vdev_ops, ==, &vdev_draid_ops);
+	ASSERT(cfg != NULL);
+	ASSERT3U(vd->vdev_nparity, ==, cfg->dcf_parity);
+	ASSERT3U(vd->vdev_children, ==, cfg->dcf_children);
+	ASSERT(cfg->dcf_zero_abd != NULL);
+}
+
+uint64_t
+vdev_draid_get_groupsz(const vdev_t *vd, boolean_t mirror)
+{
+	struct vdev_draid_configuration *cfg = vd->vdev_tsd;
+	uint64_t copies;
+
+	vdev_draid_assert_vd(vd);
+
+	copies = mirror ?
+	    vd->vdev_nparity + 1 : vd->vdev_nparity + cfg->dcf_data;
+	return (copies << DRAID_SLICESHIFT);
+}
+
+#define	DRAID_PERM_ASIZE(vd) (((vd)->vdev_children - \
+	((struct vdev_draid_configuration *)(vd)->vdev_tsd)->dcf_spare) \
+	<< DRAID_SLICESHIFT)
+
+uint64_t
+vdev_draid_offset2group(const vdev_t *vd, uint64_t offset, boolean_t mirror)
+{
+	uint64_t perm, perm_off, group, copies, groups_per_perm;
+	struct vdev_draid_configuration *cfg = vd->vdev_tsd;
+
+	vdev_draid_assert_vd(vd);
+
+	perm = offset / DRAID_PERM_ASIZE(vd);
+	perm_off = offset % DRAID_PERM_ASIZE(vd);
+	group = perm_off / vdev_draid_get_groupsz(vd, mirror);
+
+	copies = mirror ?
+	    vd->vdev_nparity + 1 : vd->vdev_nparity + cfg->dcf_data;
+	groups_per_perm = (vd->vdev_children - cfg->dcf_spare + copies - 1)
+	    / copies;
+
+	return (perm * groups_per_perm + group);
+}
+
+uint64_t
+vdev_draid_group2offset(const vdev_t *vd, uint64_t group, boolean_t mirror)
+{
+	struct vdev_draid_configuration *cfg = vd->vdev_tsd;
+	uint64_t copies, groups_per_perm, offset;
+
+	vdev_draid_assert_vd(vd);
+
+	copies = mirror ?
+	    vd->vdev_nparity + 1 : vd->vdev_nparity + cfg->dcf_data;
+	groups_per_perm = (vd->vdev_children - cfg->dcf_spare + copies - 1)
+	    / copies;
+
+	offset = DRAID_PERM_ASIZE(vd) * (group / groups_per_perm);
+	offset +=
+	    vdev_draid_get_groupsz(vd, mirror) * (group % groups_per_perm);
+	return (offset);
+}
+
+boolean_t
+vdev_draid_is_remainder_group(const vdev_t *vd,
+    uint64_t group, boolean_t mirror)
+{
+	struct vdev_draid_configuration *cfg = vd->vdev_tsd;
+	uint64_t copies, groups_per_perm;
+
+	vdev_draid_assert_vd(vd);
+
+	copies = mirror ?
+	    vd->vdev_nparity + 1 : vd->vdev_nparity + cfg->dcf_data;
+	groups_per_perm = (vd->vdev_children - cfg->dcf_spare + copies - 1)
```

Comment:
if the most likely case is not draid-mirror, then this calculation is better relocated below the assert below, prior to its use